### ISSUE TITLE
[xml] XMLNS prefix is not pre-declared in xpath query

### ISSUE THREAD

**Issue by guxiaonian** [pmd/pmd#2766]
Hello there:
I want to locate the xmlns:android location of the following XML file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
         package="com.a.b">

        <application
             android:allowBackup="true"
             android:icon="@mipmap/ic_launcher"
             android:label="@string/app_name"
             android:roundIcon="@mipmap/ic_launcher_round"
             android:supportsRtl="true"
             android:theme="@style/AppTheme">
         <activity android:name=".MainActivity">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />

                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
     </application>

</manifest>
```

My current approach is:

`//manifest[contains(@xmlns:android,'http://schemas.android.com/apk/res/android')]`

But there is an error in Xpath, can you tell me the correct way?

**Comment by oowekyala** [comment]
Is the error something like "Prefix `xmlns` has not been declared" ? In that case, maybe it's because we need to declare this namespace prefix in saxon (it would be a bug on our part)... In XPath 3.1, I think the prefixes xml and xmlns are special-cased in the spec, so are kind-of declared implicitly ([source](https://www.w3.org/TR/2017/REC-xpath-31-20170321/#id-consistency-constraints)). But in XPath 2.0 there's no special case for those and we would need to declare them manually in XPathRule

But in the XML document you provided, there's no `manifest` element with an `xmlns:android` attribute... Is this normal?

**Comment by adangel** [comment]
> But in the XML document you provided, there's no `manifest` element with an `xmlns:android` attribute... Is this normal?

Actually, there is: I just edited the original post to add syntax highlighting using markdown and now it appears (and is not rendered by the browser anymore...)

I'm not sure, we are handling namespacing at all.... one would need to declare the namespace prefix and we would use that to parse the XML document and execute the XPath. However, I'm not sure, if you can search for elements, which declare a namespace.... you would be able to search for nodes, that use the namespace by defining a attribute using that namespace....

**Comment by oowekyala** [comment]
This will be fixed by #3864. While `XPathRule` does not consider namespaces at all, the new rule `DomXPathRule` allows you to write
```
/manifest[namespace-uri-for-prefix('android', .) = 'http://schemas.android.com/apk/res/android']
```
